File: agent/src/services/metrics_service.py
from datetime import datetime, timezone

from src.collectors.metrics import MetricsCollector
from src.config.settings import AgentSettings
from src.services.http_client import HttpClient


class MetricsService:
    def __init__(
        self,
        http_client: HttpClient,
        settings: AgentSettings,
        metrics_collector: MetricsCollector,
    ):
        self.http_client = http_client
        self.settings = settings
        self.metrics_collector = metrics_collector

    def collect_and_send(self) -> None:
        metrics = self.metrics_collector.collect()

        payload = {
            "agentId": str(self.settings.agent_id),
            "cpuPercent": metrics.cpu_percent,
            "memoryUsed": metrics.memory_used,
            "memoryTotal": metrics.memory_total,
            "diskUsed": metrics.disk_used,
            "diskTotal": metrics.disk_total,
            "netBytesSent": metrics.net_bytes_sent,
            "netBytesRecv": metrics.net_bytes_recv,
            "collectedAt": datetime.now(timezone.utc).isoformat(),
        }

        self.http_client.post(
            "/api/v1/metrics",
            payload,
        )

File: frontend/src/services/metricsService.ts

import { apiClient } from "./apiClient";

export interface Metric {
  id: number;
  endpointId: number;
  cpuPercent: number | null;
  memoryUsed: number | null;
  memoryTotal: number | null;
  diskUsed: number | null;
  diskTotal: number | null;
  netBytesSent: string | null;
  netBytesRecv: string | null;
  collectedAt: string;
}

export async function getMetrics(
  endpointId: number,
  limit = 20,
): Promise<Metric[]> {
  return apiClient.get<Metric[]>(
    `/api/v1/metrics?endpointId=${endpointId}&limit=${limit}`,
  );
}